File: packages/sdk-java/src/main/java/com/zenvio/sdk/Zenvio.java
package com.zenvio.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 * Cliente Zenvio — WhatsApp, SMS, Email e envio por template (messages).
 * Em 4xx/5xx lança {@link ZenvioApiException}.
 */
public class Zenvio {
    private final String apiKey;
    private final String baseUrl;
    private final HttpClient httpClient;
    final ObjectMapper objectMapper;

    public final WhatsAppNamespace whatsapp;
    public final SmsNamespace sms;
    public final EmailNamespace email;
    public final MessagesNamespace messages;

    public Zenvio(String apiKey) {
        this(apiKey, "https://api.zenvio.com/v1");
    }

    public Zenvio(String apiKey, String baseUrl) {
        this(apiKey, baseUrl, HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build());
    }

    public Zenvio(String apiKey, String baseUrl, HttpClient httpClient) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.httpClient = httpClient;
        this.objectMapper = new ObjectMapper();
        this.whatsapp = new WhatsAppNamespace(this);
        this.sms = new SmsNamespace(this);
        this.email = new EmailNamespace(this);
        this.messages = new MessagesNamespace(this);
    }

    /**
     * Executa uma requisição e retorna o body parseado. Em status >= 400 lança ZenvioApiException.
     */
    @SuppressWarnings("unchecked")
    protected <T> T request(String method, String path, Object body, Class<T> responseType) {
        try {
            String jsonBody = body != null ? objectMapper.writeValueAsString(body) : "";

            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Authorization", "Bearer " + apiKey)
                    .header("Content-Type", "application/json")
                    .header("User-Agent", "Zenvio-Java-SDK/0.2.0");

            if ("GET".equals(method) || "DELETE".equals(method)) {
                builder.method(method, HttpRequest.BodyPublishers.noBody());
            } else {
                builder.method(method, HttpRequest.BodyPublishers.ofString(jsonBody));
            }

            HttpRequest request = builder.build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() >= 400) {
                throw new ZenvioApiException(response.statusCode(), response.body());
            }

            if (responseType == Void.class || response.body() == null || response.body().isEmpty()) {
                return null;
            }
            return objectMapper.readValue(response.body(), responseType);
        } catch (ZenvioApiException e) {
            throw e;
        } catch (Exception e) {
            throw new RuntimeException("SDK request failed: " + e.getMessage(), e);
        }
    }
}

File: packages/sdk-php/src/Zenvio.php
<?php

namespace Zenvio;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;
use Zenvio\Exception\ZenvioApiException;

class Zenvio
{
    private string $apiKey;
    private string $baseUrl;
    private GuzzleClient $httpClient;

    public Whatsapp $whatsapp;
    public Sms $sms;
    public Email $email;
    public Messages $messages;

    public function __construct(string $apiKey, string $baseUrl = 'https://api.zenvio.com/v1', array $guzzleOptions = [])
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = rtrim($baseUrl, '/');

        $defaultOptions = [
            'base_uri' => $this->baseUrl . '/',
            'timeout' => 30.0,
            'headers' => [
                'Authorization' => "Bearer {$this->apiKey}",
                'Content-Type' => 'application/json',
                'User-Agent' => 'Zenvio-PHP-SDK/0.2.0',
            ],
        ];

        $this->httpClient = new GuzzleClient(array_merge($defaultOptions, $guzzleOptions));
        $this->whatsapp = new Whatsapp($this);
        $this->sms = new Sms($this);
        $this->email = new Email($this);
        $this->messages = new Messages($this);
    }

    /**
     * Executa a requisição HTTP. Em status >= 400 lança ZenvioApiException.
     * @return array decoded JSON (vazio ou null em 204 sem body)
     * @throws ZenvioApiException
     */
    public function request(string $method, string $path, ?array $body = null): array
    {
        $path = ltrim($path, '/');
        $options = [];

        if ($body !== null && $body !== [] && !in_array($method, ['GET', 'DELETE'], true)) {
            $options['json'] = $body;
        }

        try {
            $response = $this->httpClient->request($method, $path, $options);
        } catch (GuzzleException $e) {
            $resp = $e->getResponse();
            $code = $resp ? $resp->getStatusCode() : 0;
            $bodyContent = $resp ? $resp->getBody()->getContents() : $e->getMessage();
            throw new ZenvioApiException($code, $bodyContent, $e);
        }

        $contents = $response->getBody()->getContents();
        if ($contents === '') {
            return [];
        }

        $decoded = json_decode($contents, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new ZenvioApiException($response->getStatusCode(), $contents);
        }

        return $decoded;
    }
}
